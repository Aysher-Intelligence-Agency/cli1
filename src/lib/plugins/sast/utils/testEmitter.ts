import { emitter as codeEmitter } from '@snyk/code-client';
import { spinner } from '../../../spinner';
import * as debugLib from 'debug';
import { MAX_STRING_LENGTH } from '../../../constants';

export function analysisProgressUpdate(): void {
  let currentMessage = '';
  const showSpinner = (message: string): Promise<void> | undefined => {
    if (currentMessage === message) return;

    spinner.clear<void>(currentMessage)();
    currentMessage = message;
    return spinner(message);
  };

  codeEmitter.on('supportedFilesLoaded', () =>
    showSpinner(`Supported extensions loaded`),
  );
  codeEmitter.on('scanFilesProgress', (processed: number) =>
    showSpinner(`Scanning files: ${Math.round(processed / 100)}00`),
  );
  codeEmitter.on('createBundleProgress', (processed: number, total: number) =>
    showSpinner(`Batching file upload: ${processed} / ${total}`),
  );
  codeEmitter.on('uploadBundleProgress', (processed: number, total: number) =>
    showSpinner(`Upload progress: ${processed} / ${total}`),
  );
  codeEmitter.on('analyseProgress', (data: any) =>
    showSpinner(`${data.status}: ${Math.round(data.progress * 100)}%`),
  );
  codeEmitter.on('sendError', (error) => {
    throw error;
  });
  codeEmitter.on('apiRequestLog', (data) => {
    const debug = debugLib('snyk-code');
    if (data.length > MAX_STRING_LENGTH) {
      // limit the string length as truncation doesn't always happen, causing the CLI to end unexpectedly
      debug(
        '---> API request log ',
        data.slice(0, MAX_STRING_LENGTH) + '...(log line truncated)',
      );
    } else {
      debug('---> API request log ', data);
    }
  });
}
