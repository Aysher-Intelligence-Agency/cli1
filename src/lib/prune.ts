import * as _debug from 'debug';
import { DepGraph, legacy } from '@snyk/dep-graph';

import { DepTree } from './types';
import config from './config';
import { TooManyVulnPaths } from './errors';
import * as analytics from '../lib/analytics';
import { SupportedPackageManagers } from './package-managers';

const debug = _debug('snyk:prune');

const { depTreeToGraph, graphToDepTree } = legacy;

export async function pruneGraph(
  depGraph: DepGraph,
  packageManager: SupportedPackageManagers,
  pruneIsRequired = false,
): Promise<DepGraph> {
  const prune = shouldPrune(depGraph, pruneIsRequired);

  debug('rootPkg', depGraph.rootPkg);
  debug('shouldPrune: ' + prune);
  if (prune) {
    debug('Trying to prune the graph');
    const pruneStartTime = Date.now();
    const prunedTree = (await graphToDepTree(depGraph, packageManager, {
      deduplicateWithinTopLevelDeps: true,
    })) as DepTree;
    const graphToTreeEndTime = Date.now();
    analytics.add(
      'prune.graphToTreeDuration',
      graphToTreeEndTime - pruneStartTime,
    );
    const prunedGraph = await depTreeToGraph(prunedTree, packageManager);
    analytics.add('prune.treeToGraphDuration', Date.now() - graphToTreeEndTime);

    const { completed: postPrunePathsCountCompleted } =
      countPathsToRootEarlyExit(prunedGraph, config.MAX_PATH_COUNT);

    if (!postPrunePathsCountCompleted) {
      debug('Too many paths to process the project');
      //TODO replace the throw below with TooManyPaths we do not calculate vuln paths there
      throw new TooManyVulnPaths();
    }
    return prunedGraph;
  }
  return depGraph;
}

function shouldPrune(depGraph: DepGraph, pruneIsRequired: boolean): boolean {
  // Short circuit is possible if this flag is set
  if (pruneIsRequired) {
    return true;
  }

  const { completed } = countPathsToRootEarlyExit(
    depGraph,
    config.PRUNE_DEPS_THRESHOLD,
  );

  return !completed;
}

function countPathsToRootEarlyExit(
  depGraph: DepGraph,
  limit: number,
): { count: number; completed: boolean } {
  let accumulatedPathCount = 0;
  for (const pkg of depGraph.getPkgs()) {
    accumulatedPathCount += depGraph.countPathsToRoot(pkg, { limit });

    if (accumulatedPathCount > limit) {
      return { completed: false, count: accumulatedPathCount };
    }
  }

  return { completed: true, count: accumulatedPathCount };
}
