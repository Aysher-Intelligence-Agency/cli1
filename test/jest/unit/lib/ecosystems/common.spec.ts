import { Writable } from 'stream';

import { isUnmanagedEcosystem } from '../../../../../src/lib/ecosystems/common';
import { handleProcessingStatus } from '../../../../../src/lib/polling/common';
import { FailedToRunTestError } from '../../../../../src/lib/errors';
import { printUnmanagedDepGraph } from '../../../../../src/lib/ecosystems/test';
import * as utils from '../../../../../src/lib/ecosystems/unmanaged/utils';
import { DepGraphDataOpenAPI } from '../../../../../src/lib/ecosystems/unmanaged/types';

describe('isUnmanagedEcosystem fn', () => {
  it.each`
    actual      | expected
    ${'cpp'}    | ${true}
    ${'docker'} | ${false}
    ${'code'}   | ${false}
  `(
    'should validate that given $actual as input, is considered or not an unmanaged ecosystem',
    ({ actual, expected }) => {
      expect(isUnmanagedEcosystem(actual)).toEqual(expected);
    },
  );
});

describe('handleProcessingStatus fn', () => {
  it.each`
    actual         | expected
    ${'CANCELLED'} | ${'Failed to process the project. Please run the command again with the `-d` flag and contact support@snyk.io with the contents of the output'}
    ${'ERROR'}     | ${'Failed to process the project. Please run the command again with the `-d` flag and contact support@snyk.io with the contents of the output'}
  `(
    'should validate that given $actual as input, is considered or not an unmanaged ecosystem',
    ({ actual, expected }) => {
      expect(() => {
        handleProcessingStatus({ status: actual } as any);
      }).toThrowError(new FailedToRunTestError(expected));
    },
  );
});

describe('printUnmanagedDepGraph fn', () => {
  it('should print the dep-graph', async () => {
    const mockedUnmanagedDepGraph: DepGraphDataOpenAPI = {
      schema_version: '1.2.0',
      pkg_manager: {
        name: 'cpp',
      },
      pkgs: [
        {
          id: 'root-node@0.0.0',
          info: {
            name: 'root-node',
            version: '0.0.0',
          },
        },
        {
          id: 'https://ftp.gnu.org|cpio@2.12',
          info: {
            name: 'https://ftp.gnu.org|cpio',
            version: '2.12',
          },
        },
      ],
      graph: {
        root_node_id: 'root-node',
        nodes: [
          {
            node_id: 'root-node',
            pkg_id: 'root-node@0.0.0',
            deps: [
              {
                node_id: 'https://ftp.gnu.org|cpio@2.12',
              },
            ],
          },
          {
            node_id: 'https://ftp.gnu.org|cpio@2.12',
            pkg_id: 'https://ftp.gnu.org|cpio@2.12',
            deps: [],
          },
        ],
      },
    };
    jest
      .spyOn(utils, 'getUnmanagedDepGraph')
      .mockImplementation(() => Promise.resolve([mockedUnmanagedDepGraph]));
    let buffer = Buffer.alloc(0);
    const mockDest = new Writable({
      write(chunk, _, next) {
        buffer = Buffer.concat([buffer, chunk]);
        next();
      },
    });

    const { result } = await printUnmanagedDepGraph({}, 'foo/bar', mockDest);

    expect(result).toBe('');
    expect(buffer.toString()).toMatchSnapshot();
  });
});
