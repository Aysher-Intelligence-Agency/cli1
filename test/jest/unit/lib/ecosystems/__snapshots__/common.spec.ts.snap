// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printUnmanagedDepGraph fn should print the dep-graph 1`] = `
"DepGraph data:
{"schemaVersion":"1.2.0","pkgManager":{"name":"cpp"},"pkgs":[{"id":"root-node@0.0.0","info":{"name":"root-node","version":"0.0.0"}},{"id":"https://ftp.gnu.org|cpio@2.12","info":{"name":"https://ftp.gnu.org|cpio","version":"2.12"}}],"graph":{"rootNodeId":"root-node","nodes":[{"nodeId":"root-node","pkgId":"root-node@0.0.0","deps":[{"nodeId":"https://ftp.gnu.org|cpio@2.12"}]},{"nodeId":"https://ftp.gnu.org|cpio@2.12","pkgId":"https://ftp.gnu.org|cpio@2.12","deps":[]}]}}
DepGraph target:
foo/bar
DepGraph end

"
`;
